resolver/addressable_resolver_test.go

```diff
@@ -628,6 +649,232 @@ Namespace: a lowercase RFC 1123 label must consist of lower case alphanumeric ch
+		customResolvers []resolver.RefResolverFunc
+		wantAddress     *duckv1.Addressable
+		wantErr         string
+	}{"address and addresses are not set": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRef(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableNoAddresses(),
+		},
+		addr:    addressableNoAddresses(),
+		wantErr: fmt.Sprintf("address not set for %s", addressableKnativeRef()),
+	}, "ref.address is set on destination and is present in target addressable addresses": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRefWithAddress(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithAddresses(),
+		},
+		wantAddress: addrWithNameAndCACerts(),
+		addr:        addressableWithAddresses(),
+	}, "ref.address is set on destination and is NOT present in target addressable addresses": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRefWithAddress(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithDifferentAddresses(),
+		},
+		addr:    addressableWithDifferentAddresses(),
+		wantErr: fmt.Sprintf("address with name %q not found for %s", *addressableKnativeRefWithAddress().Address, addressableKnativeRefWithAddress()),
+	}, "address and addresses are set but no destination.ref.address is set": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRef(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithAddresses(),
+		},
+		addr:        addressableWithAddresses(),
+		wantAddress: addrWithNameAndCACerts(),
+	}, "only address is set and no destination.ref.address is set": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRef(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithAddressOnly(),
+		},
+		addr:        addressableWithAddressOnly(),
+		wantAddress: addrWithNameAndCACerts(),
+	}}
+
+	for n, tc := range tests {
+		t.Run(n, func(t *testing.T) {
+			ctx, _ := fakedynamicclient.With(context.Background(), scheme.Scheme, tc.objects...)
+			ctx = addressable.WithDuck(ctx)
+			r := resolver.NewURIResolverFromTracker(ctx, tracker.New(func(types.NamespacedName) {}, 0), tc.customResolvers...)
+
+			// Run it twice since this should be idempotent. URI Resolver should
+			// not modify the cache's copy.
+			_, _ = r.AddressableFromDestinationV1(ctx, tc.dest, tc.addr)
+			addr, gotErr := r.AddressableFromDestinationV1(ctx, tc.dest, tc.addr)
+
+			if gotErr != nil {
+				if tc.wantErr != "" {
+					if got, want := gotErr.Error(), tc.wantErr; got != want {
+						t.Errorf("Unexpected error (-want, +got) =\n%s", cmp.Diff(want, got))
+					}
+				} else {
+					t.Error("Unexpected error:", gotErr)
+				}
+				return
+			}
+			if got, want := *addr.Name, *tc.wantAddress.Name; got != want {
+				t.Errorf("Unexpected object (-want, +got) =\n%s", cmp.Diff(want, got))
+			}
```

Comment:
For this I'd use this condition:
```suggestion
			if !equality.Semantic.DeepEqual(addr, tc.wantAddress) {
				t.Errorf("Unexpected object (-want, +got) =\n%s", cmp.Diff(tc.wantAddress, addr))
			}
```

The equality package is: `"k8s.io/apimachinery/pkg/api/equality"`